### ISSUE TITLE
Trim responses to not return too much information

### ISSUE THREAD

**Issue by samredai** [DataJunction/dj#328]
For the GET "/nodes/" endpoint, we return a large payload of details for every node. I'm not sure how useful this would be in common usage of the API and it might be better to return less such as a simple List[str] of node names that can be plugged into the GET "/nodes/{name}/" endpoint. Alternatively, we could have the response be a dictionary with a key for each node type and the value a list of names of nodes of that type.

We could always maintain the current route behavior in a GET "/nodes/details/" route or something similar.

**Comment by agorajek** [comment]
+1 for the no-argument option to just return a list of node names, wo details. 

Is there a REST api standard to pass a list of node names and get the them back with details, as a generalization of a single node query?

**Comment by shangyian** [comment]
Realistically, given enough nodes, wouldn't we need to set up pagination? I also think we should add some filters to the GET /nodes/ endpoint -- for example, let users provide a query arg `type` to filter down to nodes of a specific type.

**Comment by shangyian** [comment]
Yeah, that's the idea. Although on second thought, the GraphQL API pretty much does exactly this. In fact, given that we will have a GraphQL API, maybe we shouldn't worry too much about large payloads for the REST API.

**Comment by samredai** [comment]
Coming back to this thread as I've recently been running some performance tests. A relatively small DJ cluster (3 containers, 8cpus, 16g) performs well with 100 simulated users hammering the service at ~50 requests per second. But the biggest impact to the average response time are the endpoints that return long lists of nodes with all of the related metadata, nested all the way into the node revision and the node revision's details. As the number of nodes gets into the thousands, there's significant spikes in latency when all users bum rush the `/nodes/` or `/namespaces/{namespace}/` endpoints. In the chart below you can see that the first user startup requests have very low latency (creating a new namespace, creating 26 nodes, 10 dimension links -> median of ~600ms, average of ~800ms).

But once all 100 users start hitting the listing endpoints, each request returning full details on 2600 nodes (100x26), you can see the significant latency spikes (yellow) even though the total requests per second is dramatically lower (green).

<img width="1703" alt="Screenshot 2023-04-22 at 5 16 23 PM" src="https://user-images.githubusercontent.com/43911210/233806938-05b0bbb7-11c7-400a-b729-1eba9892f2bf.png">

Horizontally scaling does solve this. Adding more containers and scaling the database brings the latency down to pretty reasonable levels. However there's definitely a huge amount of optimization we can do here since we return so much information when the user just wants to see a list of node names in the system, probably satisfied with just a response schema of `["string"]`.

I'd like to propose the following:

- [ ] For the [/namespace/{namespace}/](https://github.com/DataJunction/dj/blob/main/dj/api/namespaces.py#L77) endpoint, we limit it to just a super simple select of `name`, `type`, and `description` fields for active node revisions.
- [ ] We remove the [DJClient.nodes](https://github.com/DataJunction/dj/blob/main/client/python/datajunction/client.py#L309-L315) method from the client (for now we can keep the `/nodes/` endpoint on the server for server admins to use, but it should probably be removed eventually and server admins can run the same query directly against the database when needed)
- [ ] We remove all of the type specific node listing methods on `DJClient` such as [DJClient.sources](https://github.com/DataJunction/dj/blob/main/client/python/datajunction/client.py#L89-L97), directing users to use the equivalent methods on the [Namespace](https://github.com/DataJunction/dj/blob/main/client/python/datajunction/client.py#L790-L854) class.
- [ ] We move the [DJClient.get_nodes_in_namespace](https://github.com/DataJunction/dj/blob/main/client/python/datajunction/client.py#L405-L415) helper method to live directly on the `Namespace` class and we change the logic to first call `/namespace/{namespace}/` to get a list of nodes in the namespace, then iterate over that list and make individual calls to `/nodes/{node_name}/`. Even though that adds some overhead compared to a single call, it leverages the benefits of horizontal scaling. The requests will be spread across different containers on the server side and different database nodes if a distributed database is used.

Curious to hear everyone's thoughts and if anyone is interested in working on this or splitting some of these tasks, let me know!

**Comment by shangyian** [comment]
@samredai thanks for digging into this! I can go ahead and make the changes to the client -- I was working on adding some more comprehensive tests, so might as well do this as a part of that. 

I also think that we should just go ahead and simplify the `/nodes/` endpoint to return a list of strings, and if we still want to replicate the existing `/nodes/`, we can move that to a different endpoint. Currently only the react-flow UI is dependent on the existing `/nodes/` endpoint, and that can always be modified (and would probably be more performant as well). 

Of course, this might be a moot point with the GraphQL interface, but I think it's still a bit early to start work on that.

**Comment by CircArgs** [comment]
Happy to put together the graphql API again once the rest API is stabilized

**Comment by samredai** [comment]
> I also think that we should just go ahead and simplify the `/nodes/` endpoint to return a list of strings, and if we still want to replicate the existing `/nodes/`, we can move that to a different endpoint. Currently only the react-flow UI is dependent on the existing `/nodes/` endpoint, and that can always be modified (and would probably be more performant as well).

Sounds good! If the structure of the input to the react flow library is generic enough, maybe we can have a dedicated `/dag/` endpoint that produces the entire structure server side. It'll only need to contain the information that feels appropriate to show in the visual dag.

> Happy to put together the graphql API again once the rest API is stabilized

Nice! I think we're very close to stabilizing the API for the first official release. This issue is probably one of the last modifications we'll need to make to the API. We should all have a discussion soon on any last changes we want to make before freezing the API spec for `0.*` and keeping it frozen until `1.0.0`.
